Placeholder for Blog for Service CIDR (#49989)
kubernetes/website: content/en/docs/tasks/network/extend-service-ip-ranges.md

```
Error from server (NotFound): servicecidrs.networking.k8s.io "newcidr1" not found
```

## Kubernetes Service CIDR Policies

Cluster administrators can implement policies to control the creation and
modification of ServiceCIDR resources within the cluster. This allows for
centralized management of the IP address ranges used for Services and helps
prevent unintended or conflicting configurations. Kubernetes provides mechanisms
like Validating Admission Policies to enforce these rules.

### Preventing Unauthorized ServiceCIDR Creation/Update using Validating Admission Policy

There can be situations that the cluster administrators want to restrict the
ranges that can be allowed or to completely deny any changes to the cluster
Service IP ranges.

> {{< note >}}
> The default "kubernetes" ServiceCIDR is created by the kube-apiserver
> to provide consistency in the cluster and is required for the cluster to work,
> so it always must be allowed.
> {{</ note >}}

Reviewer: OK, so re-proofreading the examples below, I got momentarily confused by the switch from `object.metadata.name != 'kubernetes'` in the first example, to `object.metadata.name == 'kubernetes'` in the second... How about if we just suggest that you should always use `matchConditions` to ignore the default ServiceCIDR, and then rewrite the "servicecidrs.deny" policy to use that as well?

````suggestion
{{< note >}}
The default "kubernetes" ServiceCIDR is created by the kube-apiserver
to provide consistency in the cluster and is required for the cluster to work,
so it always must be allowed. You can ensure your `ValidatingAdmissionPolicy`
doesn't restrict the default ServiceCIDR by adding the clause:

```yaml
  matchConditions:
  - name: 'exclude-default-servicecidr'
    expression: "object.metadata.name != 'kubernetes'"
```

as in the examples below.
{{</ note >}}
````

the value of `allowed` to something appropriate for you cluster.

...

Consult the [CEL documentation](https://kubernetes.io/docs/reference/using-api/cel/)
to learn more about CEL if you want to write your own validation `expression`.

```yaml
apiVersion: admissionregistration.k8s.io/v1
kind: ValidatingAdmissionPolicy
metadata:
  name: "servicecidrs.default"
spec:
  failurePolicy: Fail
  matchConstraints:
    resourceRules:
    - apiGroups:   ["networking.k8s.io"]
      apiVersions: ["v1","v1beta1"]
      operations:  ["CREATE", "UPDATE"]
      resources:   ["servicecidrs"]
  matchConditions:
  - name: 'exclude-default-servicecidr'
    expression: "object.metadata.name != 'kubernetes'"
  variables:
  - name: allowed
    expression: "['10.96.0.0/16','2001:db8::/64']"
  validations:
  - expression: "object.spec.cidrs.all(currentCIDR, variables.allowed.exists(allowedCIDR, cidr(allowedCIDR).containsCIDR(currentCIDR)))"
  # For all CIDRs (i) listed in the spec.cidrs of the submitted ServiceCIDR
  # object, check if there exists at least one CIDR (j) in the predefined
  # allowed list such that the allowed CIDR (j) fully contains the submitted

>   # CIDR (i)."

Reviewer: Comment still refers to the old variable names...

Also, maybe "newCIDR" instead of "currentCIDR"? And maybe "in the `allowed` list of the VAP" rather than "in the predefined allowed list"? Otherwise, yeah, I think this is the best explanation you can give that will fit in a comment.

> The following is an example of a `ValidatingAdmissionPolicy` that only allows
> ServiceCIDRs to be created if they are subranges of the given `allowed` ranges.
> (So the example policy would allow a ServiceCIDR with `cidrs: ['10.96.1.0/24']`
> or `cidrs: ['2001:db8:0:0:ffff::/80', '10.96.0.0/20']` but would not allow a
> ServiceCIDR with `cidrs: ['172.20.0.0/16']`.) You can copy this policy and change
> the value of `allowed` to something appropriate for you cluster.
> 
> ...
> 
> Consult the [CEL documentation](https://kubernetes.io/docs/reference/using-api/cel/)
> to learn more about CEL if you want to write your own validation `expression`.
> 
> ```yaml
> apiVersion: admissionregistration.k8s.io/v1
> kind: ValidatingAdmissionPolicy
> metadata:
>   name: "servicecidrs.default"
> spec:
>   failurePolicy: Fail
>   matchConstraints:
>     resourceRules:
>     - apiGroups:   ["networking.k8s.io"]
>       apiVersions: ["v1","v1beta1"]
>       operations:  ["CREATE", "UPDATE"]
>       resources:   ["servicecidrs"]
>   matchConditions:
>   - name: 'exclude-default-servicecidr'
>     expression: "object.metadata.name != 'kubernetes'"
>   variables:
>   - name: allowed
>     expression: "['10.96.0.0/16','2001:db8::/64']"
>   validations:
>   - expression: "object.spec.cidrs.all(currentCIDR, variables.allowed.exists(allowedCIDR, cidr(allowedCIDR).containsCIDR(currentCIDR)))"
>   # For all CIDRs (i) listed in the spec.cidrs of the submitted ServiceCIDR
>   # object, check if there exists at least one CIDR (j) in the predefined
>   # allowed list such that the allowed CIDR (j) fully contains the submitted
>   # CIDR (i)."
> ---
> apiVersion: admissionregistration.k8s.io/v1
> kind: ValidatingAdmissionPolicyBinding
> metadata:
>   name: "servicecidrs-binding"
> spec:
>   policyName: "servicecidrs.default"
>   validationActions: [Deny,Audit]
> ```

Reviewer: oh, sorry, I meant for the yaml example to go where I put the `...` and the "consult the cel documentation" to come after the example. ("Well then why didn't you use the 'Add a suggestion' button correctly, Dan?" yeah, yeah, sue me.)

Also, the comment on the CEL expression still refers to the old variable names... Also, maybe "newCIDR" instead of "currentCIDR"? And maybe "in the allowed list of the VAP" rather than "in the predefined allowed list"? Otherwise, yeah, I think this is the best explanation you can give that will fit in a comment.

````suggestion
The following is an example of a `ValidatingAdmissionPolicy` that only allows
ServiceCIDRs to be created if they are subranges of the given `allowed` ranges.
(So the example policy would allow a ServiceCIDR with `cidrs: ['10.96.1.0/24']`
or `cidrs: ['2001:db8:0:0:ffff::/80', '10.96.0.0/20']` but would not allow a
ServiceCIDR with `cidrs: ['172.20.0.0/16']`.) You can copy this policy and change
the value of `allowed` to something appropriate for you cluster.

```yaml
apiVersion: admissionregistration.k8s.io/v1
kind: ValidatingAdmissionPolicy
metadata:
  name: "servicecidrs.default"
spec:
  failurePolicy: Fail
  matchConstraints:
    resourceRules:
    - apiGroups:   ["networking.k8s.io"]
      apiVersions: ["v1","v1beta1"]
      operations:  ["CREATE", "UPDATE"]
      resources:   ["servicecidrs"]
  matchConditions:
  - name: 'exclude-default-servicecidr'
    expression: "object.metadata.name != 'kubernetes'"
  variables:
  - name: allowed
    expression: "['10.96.0.0/16','2001:db8::/64']"
  validations:
  - expression: "object.spec.cidrs.all(newCIDR, variables.allowed.exists(allowedCIDR, cidr(allowedCIDR).containsCIDR(newCIDR)))"
  # For all CIDRs (newCIDR) listed in the spec.cidrs of the submitted ServiceCIDR
  # object, check if there exists at least one CIDR (allowedCIDR) in the `allowed`
  # list of the VAP such that the allowedCIDR fully contains the newCIDR.
---
apiVersion: admissionregistration.k8s.io/v1
kind: ValidatingAdmissionPolicyBinding
metadata:
  name: "servicecidrs-binding"
spec:
  policyName: "servicecidrs.default"
  validationActions: [Deny,Audit]
```

Consult the [CEL documentation](https://kubernetes.io/docs/reference/using-api/cel/)
to learn more about CEL if you want to write your own validation `expression`.
````

#### Restrict any usage of the ServiceCIDR API

The following example demonstrates how to use a `ValidatingAdmissionPolicy` and
its binding to restrict the creation of any new Service CIDR ranges, excluding the default "kubernetes" ServiceCIDR:

```yaml
apiVersion: admissionregistration.k8s.io/v1
kind: ValidatingAdmissionPolicy
metadata:
  name: "servicecidrs.deny"
spec:
  failurePolicy: Fail
  matchConstraints:
    resourceRules:
    - apiGroups:   ["networking.k8s.io"]
      apiVersions: ["v1","v1beta1"]
      operations:  ["CREATE", "UPDATE"]
      resources:   ["servicecidrs"]

>   validations:
>   - expression: "object.metadata.name == 'kubernetes'"

Reviewer: ```suggestion
  matchConditions:
  - name: 'exclude-default-servicecidr'
    expression: "object.metadata.name != 'kubernetes'"
```

and then you can just not have any validations and it will never validate, right? Or maybe it would be better to add an explicit

```yaml
  validations:
  - expression: false
```

?

Author: I understand your point of consistency but having a expression that says false does not looks great to me

Admission Policies will be something applied by cluster admin that will always have to tune, so I expect this to be more a guidance or examples than the authoritative VAPs, so it is not bad to have different ways of achieving things and show the flexibility of VAPs

```
Error from server (NotFound): servicecidrs.networking.k8s.io "newcidr1" not found
```

## Kubernetes Service CIDR Policies

Cluster administrators can implement policies to control the creation and
modification of ServiceCIDR resources within the cluster. This allows for
centralized management of the IP address ranges used for Services and helps
prevent unintended or conflicting configurations. Kubernetes provides mechanisms
like Validating Admission Policies to enforce these rules.

### Preventing Unauthorized ServiceCIDR Creation/Update using Validating Admission Policy

The following example demonstrates how to use a `ValidatingAdmissionPolicy` and
its binding to restrict the Service CIDR ranges that can be created or updated,

> excluding the default "kubernetes" ServiceCIDR:

Reviewer: You should explain more clearly exactly what _this_ VAP does. I think it would also be good to provide an example "don't allow any ServiceCIDR changes" VAP, since it's slightly non-obvious (you have to allow the apiserver to create the default one.)

Author: added more details and both examples

```
Error from server (NotFound): servicecidrs.networking.k8s.io "newcidr1" not found
```

## Kubernetes Service CIDR Policies

Cluster administrators can implement policies to control the creation and
modification of ServiceCIDR resources within the cluster. This allows for
centralized management of the IP address ranges used for Services and helps
prevent unintended or conflicting configurations. Kubernetes provides mechanisms
like Validating Admission Policies to enforce these rules.

### Preventing Unauthorized ServiceCIDR Creation/Update using Validating Admission Policy

There can be situations that the cluster administrators want to restrict the
ranges that can be allowed or to completely deny any changes to the cluster
Service IP ranges.

{{< note >}}
The default "kubernetes" ServiceCIDR is created by the kube-apiserver
to provide consistency in the cluster and is required for the cluster to work,
so it always must be allowed.
{{</ note >}}

#### Restrict Service CIDR ranges to some specific ranges

> The following example demonstrates how to use a `ValidatingAdmissionPolicy` and
> its binding to restrict the Service CIDR ranges that can be created or updated,
> excluding the default "kubernetes" ServiceCIDR:

Reviewer: ```suggestion
The following example demonstrates how to use a `ValidatingAdmissionPolicy` and
its binding to restrict the Service CIDR ranges that can be created or updated. It
uses the FIXME function to check that all of the CIDR values specified are
subranges of one of the two given `allowed` ranges. So it would allow a ServiceCIDR
with `cidrs: ['10.96.1.0/24']` or `cidrs: ['2001:db8:0:0:ffff::/80', '10.96.0.0/20']`
but would not allow a ServiceCIDR with `cidrs: ['172.20.0.0/16']`.
```

and then somewhere have a link to somewhere that explains how to write your own expressions

Author: ok, got it now,

Author: added

```yaml
apiVersion: admissionregistration.k8s.io/v1
kind: ValidatingAdmissionPolicy
metadata:
  name: "servicecidrs.default"
spec:
  failurePolicy: Fail
  matchConstraints:
    resourceRules:
    - apiGroups:   ["networking.k8s.io"]
      apiVersions: ["v1","v1beta1"]

>       operations:  ["CREATE", "UPDATE"]

Reviewer: does UPDATE actually matter, given that `spec.cidrs` is immutable?

Also, is the `v1beta1` necessary? If so, do we need `v1alpha1` too?

Author: I made this also for our internal use cases, we have users in prod with v1beta, I'm not aware of anyone using it in alpha 😄

Reviewer: The important question to me is not whether anyone is _currently_ using those api versions, it's whether people in the future can use those api versions to bypass VAPs. Are _all_ clusters required to say `apiVersions: ["v1", "v1beta1"]` (or `apiVersions: ["v1", "v1beta1", "v1alpha1"]`) to prevent people from bypassing the VAP by creating a ServiceCIDR with an older apiVersion? (If yes, this seems like a terrible flaw in VAP. If no, then you shouldn't need to specify `"v1beta1"` here, unless you're doing it for people running versions of k8s that don't have the `"v1"` API yet, but in that case, would the VAP even be accepted?)

Author: > it's whether people in the future can use those api versions to bypass VAPs

oh, understand, sorry,  alpha is not served so it can not be used, the apiserver does not understand it.

> Are all clusters required to say

I think this is the same as webhooks, and you can say any apiversions with the asterisk.
Other thing is also the RBAC permissions you give to people, in addition to the policies that you apply, and also betas are removed after X releases (6??) I can't remember now.

AFAIK the VAP does not validate against the existing types enabled, it just install rules that match against the request, @jpbetz is the expert

Author: UPDATE can change finalizers and annotations and labels, so I think is relevant, since ServiceCIDRs are protected by finalizers

Author: the policies will be tested here https://github.com/kubernetes/kubernetes/pull/131318/commits/c666380c04991f98fc9dbd8e1073b6bea8193bc8

      resources:   ["servicecidrs"]
  matchConditions:
  - name: 'exclude-default-servicecidr'
    expression: "object.metadata.name != 'kubernetes'"
  variables:
  - name: allowed
    expression: "['10.96.0.0/16','2001:db8::/64']"
  validations:

>   - expression: "object.spec.cidrs.all(i , variables.allowed.exists(j , cidr(j).containsCIDR(i)))"

Reviewer: I don't know the syntax here but `cidr(j)` seems like it must be wrong; there is no field or variable named `cidr` anywhere.

Author: `cidr` is a macro https://github.com/kubernetes/kubernetes/issues/130441 , I'm adding test to validate the expressions since they are super complex to read and to evaluate

Reviewer: oh, it's a typecast basically? In that case, `i` and `j` are confusing variable names; I assumed they were indexes, not the actual field values.

Reviewer: a yaml comment explaining the expression would be very good here

Author: my gemini explanation

```
Okay, let's break down that CEL expression:

cel
object.spec.cidrs.all(i , variables.allowed.exists(j , cidr(j).containsCIDR(i)))
You're right, it definitely involves iterating over lists! Here's a step-by-step analysis:

object.spec.cidrs: This accesses the cidrs field within the spec of the Kubernetes object being evaluated (which is likely a ServiceCIDR object based on the surrounding code). This field is expected to be a list of strings, where each string is a CIDR block (e.g., ["10.0.1.0/24", "2001:db8:1::/64"]).

.all(i, ...): This is a CEL macro that iterates over every element in the object.spec.cidrs list. It assigns each element (each input CIDR string) to the temporary variable i one by one. The .all macro will only return true if the condition inside it (...) evaluates to true for every single element i in the list.

variables.allowed: This accesses a pre-defined variable named allowed within the admission policy's context. In the code you provided, this variable holds a list of allowed CIDR strings (e.g., ['10.96.0.0/16','2001:db8::/64']).

.exists(j, ...): This is another CEL macro nested inside the first one. For each input CIDR i, this macro iterates over every element in the variables.allowed list. It assigns each allowed CIDR string to the temporary variable j. The .exists macro will return true if the condition inside it (...) evaluates to true for at least one element j in the variables.allowed list.

cidr(j): This uses the built-in cidr function (available in Kubernetes CEL) to parse the string j (an allowed CIDR) into a special CIDR object that understands network ranges.

.containsCIDR(i): This is a method on the CIDR object created from cidr(j). It checks if the network range represented by j completely contains the network range represented by the input CIDR i. The string i is implicitly converted to a CIDR object for this comparison.

In plain English:

The expression does the following:

"For all CIDRs (i) listed in the spec.cidrs of the submitted ServiceCIDR object, check if there exists at least one CIDR (j) in the predefined allowed list such that the allowed CIDR (j) fully contains the submitted CIDR (i)."

```
:)

Author: added

---
apiVersion: admissionregistration.k8s.io/v1
kind: ValidatingAdmissionPolicyBinding
metadata:
  name: "servicecidrs-binding"
spec:
  policyName: "servicecidrs.default"
  validationActions: [Deny,Audit]
```

> #### Restrict any Service CIDR ranges

Reviewer: maybe

```suggestion
#### Restrict any usage of the ServiceCIDR API
```

```
Error from server (NotFound): servicecidrs.networking.k8s.io "newcidr1" not found
```

## Kubernetes Service CIDR Policies

Cluster administrators can implement policies to control the creation and
modification of ServiceCIDR resources within the cluster. This allows for
centralized management of the IP address ranges used for Services and helps
prevent unintended or conflicting configurations. Kubernetes provides mechanisms
like Validating Admission Policies to enforce these rules.

### Preventing Unauthorized ServiceCIDR Creation/Update using Validating Admission Policy

There can be situations that the cluster administrators want to restrict the
ranges that can be allowed or to completely deny any changes to the cluster
Service IP ranges.

{{< note >}}
The default "kubernetes" ServiceCIDR is created by the kube-apiserver
to provide consistency in the cluster and is required for the cluster to work,
so it always must be allowed.
{{</ note >}}

#### Restrict Service CIDR ranges to some specific ranges

> The following example demonstrates how to use a `ValidatingAdmissionPolicy` and
> its binding to restrict the Service CIDR ranges that can be created or updated.
> It uses the [CEL expression](https://kubernetes.io/docs/reference/using-api/cel/) `object.spec.cidrs.all(i, variables.allowed.exists(j, cidr(j).containsCIDR(i)))`
> to check that all of the CIDR values specified in the ServiceCIDR obejct are
> subranges of one of the two defined by the variable `allowed`. So it would allow
> a ServiceCIDR with `cidrs: ['10.96.1.0/24']` or `cidrs:
> ['2001:db8:0:0:ffff::/80', '10.96.0.0/20']` but would not allow a ServiceCIDR
> with `cidrs: ['172.20.0.0/16']`.

Reviewer: The CEL syntax here is pretty confusing if you're not familiar with it (the fact that `all` and `exists` are macros that work sort of like lambdas, but without any obvious syntactic indication of that fact). On second thought, I think trying to explain how it works is probably overall a lose. We should just treat this as something for the admin to cut and paste without understanding, and then link to the CEL docs for if they want to try to create their own more complicated policy.

```suggestion
The following is an example of a `ValidatingAdmissionPolicy` that only allows
ServiceCIDRs to be created if they are subranges of the given `allowed` ranges.
(So the example policy would allow a ServiceCIDR with `cidrs: ['10.96.1.0/24']`
or `cidrs: ['2001:db8:0:0:ffff::/80', '10.96.0.0/20']` but would not allow a
ServiceCIDR with `cidrs: ['172.20.0.0/16']`.) You can copy this policy and change
the value of `allowed` to something appropriate for you cluster.

...

Consult the [CEL documentation](https://kubernetes.io/docs/reference/using-api/cel/)
to learn more about CEL if you want to write your own validation `expression`.
```

```yaml
apiVersion: admissionregistration.k8s.io/v1
kind: ValidatingAdmissionPolicy
metadata:
  name: "servicecidrs.default"
spec:
  failurePolicy: Fail
  matchConstraints:
    resourceRules:
    - apiGroups:   ["networking.k8s.io"]
      apiVersions: ["v1","v1beta1"]
      operations:  ["CREATE", "UPDATE"]
      resources:   ["servicecidrs"]
  matchConditions:
  - name: 'exclude-default-servicecidr'
    expression: "object.metadata.name != 'kubernetes'"
  variables:
  - name: allowed
    expression: "['10.96.0.0/16','2001:db8::/64']"
  validations:

>   - expression: "object.spec.cidrs.all(i , variables.allowed.exists(j , cidr(j).containsCIDR(i)))"

Reviewer: still needs better variable names. `i` and `j` are indexes, not values.
It's annoying that we can't use `cidr` for the first variable, but even `c` and `a` would be better than `i` and `j`.

Author: changed to `"object.spec.cidrs.all(currentCIDR, variables.allowed.exists(allowedCIDR, cidr(allowedCIDR).containsCIDR(currentCIDR)))"`
